ui/src/combobox.tsx

```diff
@@ -59,54 +60,108 @@ export interface Combobox {
   required?: B
 }
 
-export const
-  XCombobox = ({ model: m }: { model: Combobox }) => {
-    const
-      isMultiValued = !!m.values,
-      [text, setText] = React.useState(m.value),
-      [options, setOptions] = React.useState((m.choices || []).map((text): Fluent.IComboBoxOption => ({ key: text, text }))),
-      [selected, setSelected] = React.useState<Fluent.IComboBoxOption[]>((m.values || []).map(text => ({ key: text, text }))),
-      selectOpt = (option: Fluent.IComboBoxOption) => {
-        setSelected(items => {
-          const result = option.selected ? [...items, option] : items.filter(item => item.key !== option.key)
-          wave.args[m.name] = result.map(item => item.text)
-          return result
-        })
-      },
-      onChange = (_e: React.FormEvent<Fluent.IComboBox>, option?: Fluent.IComboBoxOption, _index?: U, value?: S) => {
-        if (!option && value) {
-          const opt: Fluent.IComboBoxOption = { key: value, text: value }
-          setOptions((prevOptions = []) => [...prevOptions, opt])
-          selectOpt({...opt, selected: true})
-        }
-        if (option && isMultiValued) {
-          selectOpt(option)
-        }
-        else {
-          const v = option?.text || value || ''
-          wave.args[m.name] = v
-          setText(v)
-        }
-        if (m.trigger) wave.push()
+export const XCombobox = ({ model: m }: { model: Combobox }) => !m.values ? <ComboboxSingleSelect model={m} /> : <ComboboxMultiSelect model={m} />
+
+const ComboboxSingleSelect = ({ model: m }: { model: Omit<Combobox, 'values'> }) => {
+  const
+    [options, setOptions] = useOptions(m.choices),
+    [selected, setSelected] = React.useState<S | null>(m.value ?? null),
+    skipUseEffectForValue = React.useRef(false),
+    onChange = (_e: React.FormEvent<IComboBox>, option?: IComboBoxOption, _index?: U, value?: S) => {
+      if (!option && value) setOptions((prevOptions = []) => [...prevOptions, { key: value, text: value }])
+      
+      const v = option?.text || value || ''
+      setSelected(v)
+      
+      wave.args[m.name] = v
+      if (m.trigger) wave.push()
+
+      // Hacky: Ensure that next dynamically change to value from Wave app will trigger the useEffect
+      m.value = v
+      // The line above will trigger the useEffect for change in m.value and we need to skip it
+      skipUseEffectForValue.current = true
```

Comment:
Is this really necessary? The useEffect will set the `setSelected(m.value!)` value to the one that is already set by `setSelected(v)` in `onChange` which means the `setSelected(m.value!)` update will be skipped by React anyway. The tests pass after I removed the `skipUseEffectForValue` but maybe there is something I am missing?

Reply:
I was skipping the `useEffect` because before I had `m.value === undefined` but we don't need to skip anymore because the `useEffect` will set it to `v`. Good catch!